ui/StatusQ/src/modelentry.cpp

```diff
@@ -0,0 +1,346 @@
+
+    m_itemRemovedFromModel = itemRemovedFromModel;
+    emit itemRemovedFromModelChanged();
+}
+
+QModelIndex ModelEntry::findIndexInRange(int start, int end, const QList<int>& roles) const
+{
+    if(!m_sourceModel || m_key.isEmpty()) return {};
+
+    auto keysForRole = m_sourceModel->roleNames().keys(m_key.toUtf8());
+
+    // no matching roles found
+    if(keysForRole.isEmpty() || (!roles.isEmpty() && !roles.contains(keysForRole.first()))) return {};
+
+    for(int i = start; i < end; i++)
+    {
+        auto index = m_sourceModel->index(i, 0);
+        auto data = index.data(keysForRole.first());
+        if(data == m_value) return index;
+    }
+
+    return {};
+}
+
+void ModelEntry::resetIndex()
+{
+    auto index = QModelIndex();
+    if(m_sourceModel) index = findIndexInRange(0, m_sourceModel->rowCount());
+
+    setIndex(index);
+}
+
+void ModelEntry::tryItemResetOrUpdate()
+{
+    if(!m_index.isValid() || !itemHasCorrectRoles())
+    {
+        resetItem();
+        return;
+    }
+
+    updateItem();
+    setAvailable(true);
+}
+
+void ModelEntry::resetItem()
+{
+    // Signal order is important here
+    if(!m_index.isValid())
+    {
+        setAvailable(false);
+    }
+
+    m_item = new QQmlPropertyMap(this);
+
+    updateItem();
+
+    if(!m_index.isValid())
+    {
+        setRoles(m_item->keys());
+    }
+
+    emit itemChanged();
+
+    if(m_index.isValid())
+    {
+        setRoles(m_item->keys());
+        setAvailable(true);
+    }
+}
+
+void ModelEntry::updateItem(const QList<int>& roles /*{}*/)
+{
+    if(!m_index.isValid() || !m_sourceModel) return;
+
+    const auto& rolesRef = roles.isEmpty() ? m_sourceModel->roleNames().keys() : roles;
+
+    for(auto role : rolesRef)
+    {
+        auto roleName = m_sourceModel->roleNames().value(role);
+        auto roleValue = m_index.data(role);
+
+        if(roleValue == m_item->value(roleName)) continue;
+
+        m_item->insert(roleName, roleValue);
+        emit m_item->valueChanged(roleName, roleValue);
+    }
+    setItemRemovedFromModel(false);
+}
+
+bool ModelEntry::itemHasCorrectRoles() const
+{
+    if(!m_sourceModel || !m_item) return false;
+
+    auto itemKeys = m_item->keys();
+    auto modelRoles = m_sourceModel->roleNames().values();
+
+    return std::all_of(modelRoles.begin(),
+                       modelRoles.end(),
+                       [itemKeys](const QByteArray& role) { return itemKeys.contains(role); }) &&
+           itemKeys.size() == modelRoles.size();
+}
+
+void ModelEntry::cacheItem()
+{
+    if(!m_cacheOnRemoval) return;
+
+    for(const auto& role : qAsConst(m_roles))
+    {
+        auto roleName = m_sourceModel->roleNames().key(role.toUtf8());
+        auto roleValue = m_index.data(roleName);
+
+        if(roleValue.canConvert<QAbstractItemModel*>())
+        {
+            m_item->insert(role,
+                           QVariant::fromValue(new SnapshotModel(*roleValue.value<QAbstractItemModel*>(), true, this)));
```

Comment:
One more thing. Probably it would be better to have `m_item` as a parent here. When `QQmlPropertyMap` is properly destroyed on reset (mentioned in other comments), it will clean up snapshots as well.